Add intake limelight

tails618 opened the pull request:
- Mostly a copy from last year

tails618 committed:
Add IntakeLimelight subsystem

Copied from 2023, some changes, not done yet.

tails618 committed:
feat: Add note detection to IntakeLimelight

This commit adds the ability for the IntakeLimelight subsystem to detect notes. The `pipeline` enum now includes a new option for note detection. The `setPipeline()` method has been updated to set the pipeline correctly for note detection. A new inner class called `NoteDetection` has been added, which represents the detected note's latency, distance from the camera, and CCW angle. The `getConePos()` method has been renamed to `getNotePos()`, and it now returns an Optional<NoteDetection> instead of Optional<ConeDetection>. The `getAngleToConeDeg()` method has also been renamed to `getAngleToNoteDeg()`.

tails618 commented:
don't review yet it's not done lol

voidSevenSevenSix committed:
more stuff from last year for apriltags

balghane commented on src/main/java/frc/robot/subsystems/intakeLimelight/IntakeLimelight.java:
```diff
@@ -0,0 +1,610 @@
+        new Translation2d(botPoseTargetSpace.getZ(), botPoseTargetSpace.getX());
+    // Rotation2d rot = Rotation2d.fromDegrees(-botPoseTargetSpace.getRotation().getY());
+
+    Rotation2d rot = Rotation2d.fromDegrees(0);
```
Let's maybe return the real rotation from the tag for this year. Last year, we knew every tag was going to be at 0 deg but this year, I'm concerned about assuming anything about how the stage is placed on the field. So we may want to use the actual rotation returned by the tag detection thingy

tails618 replied:
Changed

balghane commented on src/main/java/frc/robot/subsystems/intakeLimelight/IntakeLimelight.java:
```diff
@@ -0,0 +1,610 @@
+        || tagIdRounded == 3
+        || tagIdRounded == 6
+        || tagIdRounded == 7
+        || tagIdRounded == 8) {
```
Intake cam only cares about the six stage tags, can you update for that?

tails618 replied:
Fixed. I *could* check our math state to isolate only our alliance's stage but I don't think that's necessary; whenever we're using this, it should be obvious which tag to look at

balghane commented on src/main/java/frc/robot/subsystems/intakeLimelight/IntakeLimelight.java:
```diff
@@ -0,0 +1,610 @@
+    return bestTag;
+  }
+
+  private Transform2d getRobotToScoringLocation(Pose3d targetPoseRobotSpace) {
+    Transform2d targetFromBot = getBotFromTarget(targetPoseRobotSpace);
+    return targetFromBot;
+  }
```
Do we need this? What is it doing?

tails618 replied:
Used in `checkForTag`, which is used in `LookForTag` (which doesn't exist here yet but was used last year). It seems like it gets a transform from the robot to the tag (scoring location).

balghane commented on src/main/java/frc/robot/subsystems/intakeLimelight/IntakeLimelight.java:
```diff
@@ -0,0 +1,610 @@
+    return targetFromBot;
+  }
+
+  public Optional<Transform2d> getRobotToScoringLocation() {
+    if (getPipeline() != pipeline.PIPELINE1) {
+      setPipeline(pipeline.PIPELINE1);
+    }
+    return robotToScoringLocation;
+  }
```
Do we need this? What is it doing?

tails618 replied:
Returns the existing variable itself, set in a few different places. This function is public, whereas the above function is private.

balghane commented on src/main/java/frc/robot/subsystems/intakeLimelight/IntakeLimelight.java:
```diff
@@ -0,0 +1,610 @@
+  public double getLatencySeconds() {
+    if (getPipeline() != pipeline.PIPELINE1) {
+      setPipeline(pipeline.PIPELINE1);
+    }
```
This will be useful for either tags or notes

balghane replied:
Can you dig through the docs on whether this correctly captures the latency for both neural network and tag detection pipelines?

tails618 replied:
Don't see why it wouldn't, pretty sure everything in [this section of the LL NetworkTables API](https://docs.limelightvision.io/docs/docs-limelight/apis/complete-networktables-api#basic-targeting-data) applies to all pipelines.

balghane commented on src/main/java/frc/robot/subsystems/intakeLimelight/IntakeLimelight.java:
```diff
@@ -0,0 +1,610 @@
+  /**
+   * Create an IntakeLimelight object
+   *
+   * @param angleDegrees angle from normal in degress. Looking straight out is 0, and increasing as
+   *     the camera is tilted towards the ceiling.
```
This is describing pitch

tails618 replied:
Isn't that what it should be?

balghane replied:
Comment should say it's pitch then

tails618 replied:
Done

balghane commented on src/main/java/frc/robot/subsystems/intakeLimelight/IntakeLimelight.java:
```diff
@@ -0,0 +1,610 @@
+   */
+  public double getDistanceFromTargetMeters(double targetHeight) {
+    double angle = kCameraAngleDegrees + getOffSetY();
+    if (angle < 1 || angle > 89) {
+      return 0;
+    }
```
Why is <1 not valid here?

tails618 replied:
No idea. Removed.

balghane replied:
Doesn't look removed to me

tails618 replied:
ok well nothing's perfect (except now i've actually removed it so it is perfect)

balghane commented on src/main/java/frc/robot/subsystems/intakeLimelight/IntakeLimelight.java:
```diff
@@ -0,0 +1,610 @@
+   */
+  public IntakeLimelight(double angleDegrees, double heightMeters, double targetHeightMeters) {
+    kCameraAngleDegrees = angleDegrees;
+    kCameraHeight = heightMeters;
```
This is not fixed, since the camera is on the elevator. We should pass it in when we care about it

tails618 replied:
This is the intake Limelight, which I believe is on the conveyer (that's what it looked like in [this image](https://media.discordapp.net/attachments/673346755072491521/1201927233355124807/Screenshot_2024-01-30_102653.png?ex=65de0df7&is=65cb98f7&hm=e1026de3d70d36a230912d0a29fb25e89f532bdb3262f53603ffcebb7a844b52&=&format=webp&quality=lossless&width=1640&height=1194)), so the height should be constant (the conveyer should always be in the same place when intaking)

balghane replied:
Hmmm okay

balghane commented on src/main/java/frc/robot/subsystems/intakeLimelight/IntakeLimelight.java:
```diff
@@ -0,0 +1,610 @@
+      return 0;
+    }
+    double tanTerm = Math.tan(Math.toRadians(angle));
+    double distance = (targetHeight - kCameraHeight) / tanTerm;
```
Shouldn't this be `(targetHeight - kCameraHeight) * tanTerm;`?

tails618 replied:
No. Whiteboard math here: 
![photo](https://github.com/cachemoney8096/2024-crescendo/assets/22648713/9669ba24-490c-42c4-b5a3-bcb5c7e22973)
I can type out an explanation if it's unclear.

balghane commented on src/main/java/frc/robot/subsystems/intakeLimelight/IntakeLimelight.java:
```diff
@@ -0,0 +1,610 @@
+     *
+     * <p>x_norm * d = x_target y_norm * d = diff_height z_target * d = z_target
+     */
+    double scaling = diff_height / y_norm;
```
Potential for a divide by zero here

tails618 replied:
I don't think I grasp the math well enough to figure out how to avoid it... can I get a hint?

balghane replied:
Generally you can just check if it's within some epsilon of zero and if yes, then replace it with a very small but non-zero value (like 1e-4 or something)

tails618 replied:
Should you check if it is ACTUALLY 0, and if not replace it with a very small but non-zero value? We don't care if it's within some epsilon of 0, that won't cause a divide by 0 issue.

Fixed based on this, but if it should actually be checking if it's within some epsilon I'll change it

balghane replied:
Even if it's 1e-7 that's bad

tails618 replied:
What do you mean? Even if `y_norm` is 1e-7? I don't see why that would be an issue, it should divide fine (it'll mean `scaling` is giant, but it would also be giant if I did 1e-4 like you suggested earlier).

balghane commented on src/main/java/frc/robot/subsystems/intakeLimelight/IntakeLimelight.java:
```diff
@@ -0,0 +1,610 @@
+    public double distanceMeters;
+
+    /** CCW Angle */
+    public double angleDeg;
```
Is this pitch or yaw

tails618 replied:
Yaw

balghane commented on src/main/java/frc/robot/subsystems/intakeLimelight/IntakeLimelight.java:
```diff
@@ -0,0 +1,610 @@
+   *     distance in meters.
+   */
+  public Optional<NoteDetection> getNotePos() {
+    if (getPipeline() != pipeline.PIPELINE0) {
```
Can we note somewhere above that 0 is Note and 1 is Tag

tails618 replied:
It's in the comment on `pipeline`

balghane commented on src/main/java/frc/robot/subsystems/intakeLimelight/IntakeLimelight.java:
```diff
@@ -0,0 +1,610 @@
+    return getTargetTranslation(kTargetHeight);
+  }
+
+  public Optional<Double> getXOfSmallestY(double[] corners) {
```
Can you add a comment that this grabs the object that is lowest in the image

tails618 replied:
Done

balghane commented on src/main/java/frc/robot/subsystems/intakeLimelight/IntakeLimelight.java:
```diff
@@ -0,0 +1,610 @@
+      setPipeline(pipeline.PIPELINE0);
+    }
+
+    double[] corners = table.getEntry("tcornxy").getDoubleArray(new double[0]);
+    Optional<Double> maybeXPixels = getXOfSmallestY(corners);
+    Optional<Double> maybeYHeightPixels = getObjectHeightPx(corners);
```
Do we need to filter out low confidence detections?

tails618 replied:
I don't see anything about that in the API; I assume/hope it can be done in the LL dashboard.

balghane replied:
ok let's remember to test that

balghane commented on src/main/java/frc/robot/subsystems/intakeLimelight/IntakeLimelight.java:
```diff
@@ -0,0 +1,610 @@
+    Optional<Double> maybeXPixels = getXOfSmallestY(corners);
+    Optional<Double> maybeYHeightPixels = getObjectHeightPx(corners);
+
+    if (!maybeXPixels.isPresent() || !maybeYHeightPixels.isPresent() || table.getEntry("tclass").getString("") != "note") {
```
`table.getEntry("tclass")` This is probably only the class for the object that the limelight selected, isn't it?

tails618 replied:
It is, but I think it's the right thing to do; there's other stuff here that's also only focused on the target (like `corners`, I think). 

We could get the JSON dump of *every* object it detects in the frame; do you think that's necessary?

balghane replied:
corners is the corners for all the objects, so hopefully there's a class for all the objects too?

balghane replied:
Can you put this on the minibot and test?

balghane replied:
Here's a model you can throw on the minibot https://www.chiefdelphi.com/t/limelight-neural-network-note-detection/449822/10?u=basel_a

tails618 replied:
Yeah, we'll test tomorrow. I'm leaving in a few minutes so I'm not gonna work on it rn.

balghane commented on src/main/java/frc/robot/subsystems/intakeLimelight/IntakeLimelight.java:
```diff
@@ -0,0 +1,610 @@
+      setPipeline(pipeline.PIPELINE0);
+    }
+
+    double[] corners = table.getEntry("tcornxy").getDoubleArray(new double[0]);
```
Does this include `robot` and `note`?

tails618 replied:
Yes, I believe if we aren't explicitly limiting based on class it will look for everything in the model (unless there's also filtering which classes it detects within the limelight console itself)

balghane replied:
Could be an issue, we only want to see notes here...

tails618 replied:
Ah, thought you were asking about that line specifically. In this function we check if it's a note on L535:

```java
if (!maybeXPixels.isPresent() || !maybeYHeightPixels.isPresent() || table.getEntry("tclass").getString(IntakeLimelightConstants.INTAKE_LIMELIGHT_NAME) != "note") {
      System.out.println("Didn't see note");
      return Optional.empty();
    }
```

balghane replied:
`table.getEntry("tclass")` this almost certainly only checks if the "limelight selected object" is a note, not the one we have selected above which could be either

balghane commented on src/main/java/frc/robot/subsystems/intakeLimelight/IntakeLimelight.java:
```diff
@@ -0,0 +1,610 @@
+    double yPixels = maybeYHeightPixels.get();
+    double normYPixels = (1 / halfResYPixels) * yPixels;
+    double viewplaneYPixels = viewplaneHeightPixels / 2.0 * normYPixels;
+    final double CONE_HEIGHT_METERS = Units.inchesToMeters(13.0);
```
Update

tails618 replied:
Done

balghane commented on src/main/java/frc/robot/subsystems/intakeLimelight/IntakeLimelight.java:
```diff
@@ -0,0 +1,610 @@
+    return Optional.of(new NoteDetection(getLatency(), coneDistanceMeters, adjustedAngleDegrees));
+  }
+
+  public Optional<Double> getAngleToNoteDeg() {
```
Do we need this if we have the above?

tails618 replied:
Nope! Removed.

balghane commented on src/main/java/frc/robot/subsystems/intakeLimelight/IntakeLimelightConstants.java:
```diff
@@ -0,0 +1,5 @@
+package frc.robot.subsystems.intakeLimelight;
+
+public class IntakeLimelightConstants {
+  
```
Camera pitch goes here probably

tails618 replied:
Done. I think it's 0.0deg, correct me if I'm wrong

balghane replied:
I don't know, didn't see it in the sheet

tails618 replied:
Nick wants us to add Limelight stuff to the sheet so he can go through and add that stuff. for know I've made it a placeholder.

tails618 committed:
One change I made during math class

tails618 committed:
Address PR comments

tails618 reviewed (commented):
Responses to comments

tails618 reviewed (commented):
Responses to comments

balghane commented on src/main/java/frc/robot/subsystems/intakeLimelight/IntakeLimelight.java:
```diff
@@ -0,0 +1,575 @@
+    double yPixels = maybeYHeightPixels.get();
+    double normYPixels = (1 / halfResYPixels) * yPixels;
+    double viewplaneYPixels = viewplaneHeightPixels / 2.0 * normYPixels;
+    final double NOTE_HEIGHT_METERS = Units.inchesToMeters(13.0);
```
A note is not 13 inches tall lol. It's just above the ground, we can call it 1 inch

tails618 replied:
What if it's balanced *perfectly* upright and all the teams agree to not move their robots so it doesn't fall?? This feels like a very realistic scenario and would result in a 14" tall note.

tails618 committed:
More changes

tails618 committed:
Refactor IntakeLimelight code and add calibration constants

- Refactored the import statements to remove unused imports and added necessary imports for IntakeCal.
- Added a new import statement for LimelightHelpers.LimelightTarget_Detector.
- Updated the pitchAngleDegrees parameter description in the constructor of IntakeLimelight class.
- Removed unnecessary comments and empty lines.
- Added a new constant LIMELIGHT_YAW in IntakeLimelightCal class.
- Added two new constants INTAKE_LIMELIGHT_PITCH_DEGREES and INTAKE_LIMELIGHT_HEIGHT_METERS in IntakeLimelightConstants class.

tails618 committed:
i forgor 💀 the note 🎵 has thickness 📏

balghane commented on src/main/java/frc/robot/subsystems/intakeLimelight/IntakeLimelight.java:
```diff
@@ -0,0 +1,579 @@
+  }
+
+  /** Gets the object lowest in the image */
+  public Optional<Double> getXOfSmallestY(double[] corners) {
```
Not using this anymore

tails618 replied:
Removed

balghane commented on src/main/java/frc/robot/subsystems/intakeLimelight/IntakeLimelight.java:
```diff
@@ -0,0 +1,579 @@
+    return Optional.of(corners[minYIdx - 1]);
+  }
+
+  public Optional<Double> getObjectHeightPx(double[] corners) {
```
Not using this

tails618 replied:
Removed

balghane commented on src/main/java/frc/robot/subsystems/intakeLimelight/IntakeLimelight.java:
```diff
@@ -0,0 +1,579 @@
+      setPipeline(pipeline.NOTE_PIPELINE);
+    }
+
+    LimelightHelpers.LimelightResults llresults =
+        LimelightHelpers.getLatestResults(IntakeLimelightConstants.INTAKE_LIMELIGHT_NAME);
```
There was something about the first time you run this, it takes a couple seconds? Something with the json parsing library. So you should make sure to do it before the match.

Not sure if doing it in the constructor works. Maybe disabledInit?

https://www.chiefdelphi.com/t/introducing-limelight-lib/425660/57?u=basel_a

tails618 replied:
Added to `disabledInit`

balghane commented on src/main/java/frc/robot/subsystems/intakeLimelight/IntakeLimelight.java:
```diff
@@ -0,0 +1,579 @@
+      }
+    }
+
+    double angleLimelightToNote =
+        IntakeLimelightConstants.INTAKE_LIMELIGHT_PITCH_DEGREES + lowestDetection.ty;
```
I think `INTAKE_LIMELIGHT_PITCH_DEGREES` is defined as 0 parallel to ground, 90 up. How is ty defined?

balghane replied:
Add units to var name also

tails618 replied:
"Y-coordinate of the center of the target in degrees. Positive-down, center-zero" so 0 is fine, and sign is also correct. (If we have a note 10deg below horizontal, that is +10 in limelight degrees, and if our limelight is angled 10 degrees up, that is +10 in robot degrees; (+10) + (+10) = +20, which I believe is what we want.

I have changed the name of the var.

balghane commented on src/main/java/frc/robot/subsystems/intakeLimelight/IntakeLimelight.java:
```diff
@@ -0,0 +1,579 @@
+
+  @Override
+  public void initSendable(SendableBuilder builder) {
+    builder.addDoubleProperty("Target Area", () -> getTargetArea(), null);
+    builder.addDoubleProperty("Skew", () -> getSkew(), null);
+    builder.addDoubleProperty("Latency", () -> getLatency(), null);
+    builder.addDoubleProperty("Distance", () -> getDistanceFromTargetMeters(), null);
+    builder.addDoubleProperty("Tx", () -> getOffSetX(), null);
+    builder.addDoubleProperty("Ty", () -> getOffSetY(), null);
+    builder.addDoubleProperty("Translation Distance", () -> m_lastDistance, null);
+    builder.addDoubleProperty("Translation X", () -> m_lastX, null);
+    builder.addDoubleProperty("Translation Y", () -> m_lastY, null);
+    builder.addBooleanProperty("Valid Target", () -> isValidTarget(), null);
```
We might want to remove most of this? Or anything we're not intending on using initially

tails618 replied:
I removed some stuff we definitely aren't using.  Left some stuff that might be useful.

balghane commented on src/main/java/frc/robot/subsystems/intakeLimelight/IntakeLimelightCal.java:
```diff
@@ -0,0 +1,5 @@
+package frc.robot.subsystems.intakeLimelight;
+
+public class IntakeLimelightCal {
+    public static final double LIMELIGHT_YAW = 0.0;
```
add units

tails618 replied:
Done

balghane commented on src/main/java/frc/robot/subsystems/intakeLimelight/IntakeLimelightConstants.java:
```diff
@@ -0,0 +1,10 @@
+import frc.robot.Constants;
+
+public class IntakeLimelightConstants {
+  public static final String INTAKE_LIMELIGHT_NAME = "intake";
```
I think the name is always `limelight-XYZ` so `limelight-intake`?

tails618 replied:
Good catch

tails618 committed:
More changes

balghane commented on src/main/java/frc/robot/subsystems/intakeLimelight/IntakeLimelight.java:
```diff
@@ -0,0 +1,511 @@
+      }
+    }
+
+    double angleLimelightToNoteDegrees =
+        IntakeLimelightConstants.INTAKE_LIMELIGHT_PITCH_DEGREES + lowestDetection.ty;
```
The way we're using this below is that it's the angle relative to field horizontal, from the limelight position. So if the result (note relative to field horizontal) is 4 deg down and the limelight is 7 deg up. Then the ty will report 11. So I think we need `ty - INTAKE_LIMELIGHT_PITCH_DEGREES `

tails618 replied:
Yup, you're right. Fixed

tails618 committed:
Minor fix

tails618 commented:
/format

github-actions[bot] committed:
Formatting fixes
